edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,812 @@
+                        .ForEach(
+                            exposedList =>
+                                exposedList.ForEach((item) => portList.Add(new V1ServicePort(item.Port, name: $"{item.Port}-{item.Protocol.ToLower()}", protocol: item.Protocol))));
+                }
+
+                // Handle HostConfig PortBindings entries
+                if (moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings != null)
+                {
+                    foreach (KeyValuePair<string, IList<DockerModels.PortBinding>> portBinding in moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings)
+                    {
+                        string[] portProtocol = portBinding.Key.Split('/');
+                        if (portProtocol.Length == 2)
+                        {
+                            if (int.TryParse(portProtocol[0], out int port) && this.ValidateProtocol(portProtocol[1], out string protocol))
+                            {
+                                // Entries in Docker portMap wants to expose a port on the host (hostPort) and map it to the container's port (port)
+                                // We interpret that as the pod wants the cluster to expose a port on a public IP (hostPort), and target it to the container's port (port)
+                                foreach (DockerModels.PortBinding hostBinding in portBinding.Value)
+                                {
+                                    if (int.TryParse(hostBinding.HostPort, out int hostPort))
+                                    {
+                                        portList.Add(new V1ServicePort(hostPort, name: $"{port}-{protocol.ToLower()}", protocol: protocol, targetPort: port));
+                                        onlyExposedPorts = false;
+                                    }
+                                    else
+                                    {
+                                        this.logger.PortBindingValue(module, portBinding.Key);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (portList.Count > 0)
+            {
+                // Selector: by module name and device name, also how we will label this puppy.
+                var objectMeta = new V1ObjectMeta(annotations: serviceAnnotations.GetOrElse(() => null), labels: labels, name: KubeUtils.SanitizeDNSValue(moduleIdentity.ModuleId));
+                // How we manage this service is dependent on the port mappings user asks for.
+                // If the user tells us to only use ClusterIP ports, we will always set the type to ClusterIP.
+                // If all we had were exposed ports, we will assume ClusterIP.
+                // If we have mapped ports, we are going to use "LoadBalancer" service Type - to tell the ingress controller we want this port exposed.
```

Comment:
Comment is out of date. if only exposed ports, we assume ClusterIP. Otherwise, we use the given value as the default service type.